File: java/graph/CriticalPath.java
package graph;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Stack;

public class CriticalPath {
    public int[] search(VertexNode[] graphs) {
        Stack<Integer> stack = new Stack<>();
        int[] vertexEnd = topologicalSort(graphs, stack);
        if (0 == vertexEnd.length) {
            return new int[0];
        }
        int[] vertexLatest = reverseTopologicalSort(graphs, vertexEnd, stack);

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < graphs.length; ++i) {
            if (vertexEnd[i] == vertexLatest[i]) {
                result.add(i);
            }
        }
        return result.stream().mapToInt(Integer::valueOf).toArray();
    }

    private int[] topologicalSort(VertexNode[] graphs, Stack<Integer> stack) {
        int[] inDegrees = new int[graphs.length];
        Arrays.fill(inDegrees, 0);
        for (VertexNode graph : graphs) {
            EdgeNode edge = graph.next;
            while (null != edge) {
                ++inDegrees[edge.index];
                edge = edge.next;
            }
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < graphs.length; ++i) {
            if (0 == inDegrees[i]) {
                queue.offer(i);
            }
        }

        int[] vertexEnd = new int[graphs.length];
        Arrays.fill(vertexEnd, 0);
        int count = 0;
        while (!queue.isEmpty()) {
            int index = queue.poll();
            stack.push(index);
            ++count;

            EdgeNode edge = graphs[index].next;
            int pivot = vertexEnd[index];
            while (null != edge) {
                int edgeIndex = edge.index;
                int edgeWeight = edge.weight;
                --inDegrees[edgeIndex];
                if (0 == inDegrees[edgeIndex]) {
                    queue.offer(edgeIndex);
                    if (vertexEnd[edgeIndex] < pivot + edgeWeight) {
                        vertexEnd[edgeIndex] = pivot + edgeWeight;
                    }
                }
                edge = edge.next;
            }
        }
        return (count == graphs.length) ? vertexEnd : new int[0];
    }

    private int[] reverseTopologicalSort(VertexNode[] graphs, int[] vertexEnd, Stack<Integer> stack) {
        int maxValue = vertexEnd[stack.pop()];
        int[] vertexLatest = new int[graphs.length];
        Arrays.fill(vertexLatest, maxValue);
        while (!stack.isEmpty()) {
            int index = stack.pop();
            EdgeNode edge = graphs[index].next;
            while (null != edge) {
                int edgeIndex = edge.index;
                int edgeWeight = edge.weight;
                if (vertexLatest[index] > vertexLatest[edgeIndex] - edgeWeight) {
                    vertexLatest[index] = vertexLatest[edgeIndex] - edgeWeight;
                }
                edge = edge.next;
            }
        }
        return vertexLatest;
    }

    public class EdgeNode {
        public int index = 0;
        public int weight = 0;
        public EdgeNode next = null;

        public EdgeNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    public class VertexNode {
        public int value = 0;
        public EdgeNode next = null;

        public VertexNode(int value) {
            this.value = value;
        }
    }
}

File: py/graph/CriticalPath.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class CriticalPath:
    def search(self, graphs: List[VertexNode]) -> List[int]:
        stack = []
        vertexEnd = self.__topologicalSort(graphs, stack)
        if not vertexEnd:
            return []
        vertexLatest = self.__reverseTopologicalSort(graphs, vertexEnd, stack)

        result = []
        for i, (value1, value2) in enumerate(zip(vertexEnd, vertexLatest)):
            if value1 == value2:
                result.append(i)
        return result

    def __topologicalSort(self, graphs: List[VertexNode], stack: List[int]) -> List[int]:
        inDegrees = [0 for _ in range(len(graphs))]
        for graph in graphs:
            edge = graph.next
            while edge:
                inDegrees[edge.index] += 1
                edge = edge.next

        queue = []
        for i, value in enumerate(inDegrees):
            if not value:
                queue.append(i)

        vertexEnd, count = [0 for _ in range(len(graphs))], 0
        while queue:
            index = queue.pop(0)
            stack.append(index)
            count += 1

            edge, pivot = graphs[index].next, vertexEnd[index]
            while edge:
                edgeIndex, edgeWeight = edge.index, edge.weight
                inDegrees[edgeIndex] -= 1
                if not inDegrees[edgeIndex]:
                    queue.append(edgeIndex)
                    if vertexEnd[edgeIndex] < pivot + edgeWeight:
                        vertexEnd[edgeIndex] = pivot + edgeWeight
                edge = edge.next
        return vertexEnd if count == len(graphs) else []

    def __reverseTopologicalSort(self, graphs: List[VertexNode], vertexEnd: List[int], stack: List[int]) -> List[int]:
        maxValue = vertexEnd[stack.pop()]
        vertexLatest = [maxValue for _ in range(len(graphs))]
        while stack:
            index = stack.pop()
            edge = graphs[index].next
            while edge:
                edgeIndex, edgeWeight = edge.index, edge.weight
                if vertexLatest[index] > vertexLatest[edgeIndex] - edgeWeight:
                    vertexLatest[index] = vertexLatest[edgeIndex] - edgeWeight
                edge = edge.next
        return vertexLatest
